### PR TITLE
Add dashboard display of the source of ignores

### PR DESCRIPTION
**Pull request by JesseWeinstein**

### CODE REVIEW

**Comment by yipdw** [comment]
Sorry for the delay in reviewing this.  I think I can check this out tonight.

**Comment by JesseWeinstein** [comment]
No worries about the delay -- life happens, and this is an addition, not a bug fix.

File: pipeline/wpull_hooks.py [hunk 1]
```diff
@@ -27,20 +27,21 @@
...
+    type='ignore',
+    source=source,
   )
 
   control.log(packet, ident, log_key)
 
-
-def maybe_log_ignore(url, pattern):
+# source is a single character identifying what caused the ignore
+def maybe_log_ignore(url, pattern, source='U'):
```
Comment (yipdw) [inline pipeline/wpull_hooks.py]: I'd prefer more than a single character with no default: 

1. All call sites of `maybe_log_ignore` can be found, so no default is necessary.
2. The extra size of longer `source`s doesn't matter in the default `igoff` mode, and `igon` is not left on for very long anyway.  (Also, ArchiveBot pipelines tend to be run on machines that have access to high-bandwidth pipes.)

File: pipeline/wpull_hooks.py [hunk 2]
```diff
@@ -101,7 +102,7 @@ def accept_url(url_info, record_info, verdict, reasons):
   pattern = settings.ignore_url_p(url, record_info)
 
   if pattern:
-    maybe_log_ignore(url, pattern)
+    maybe_log_ignore(url, pattern, 'A')
```
Comment (yipdw) [inline pipeline/wpull_hooks.py]: Instead of `'A'`, I think I'd prefer `accept_url`.

File: pipeline/wpull_hooks.py [hunk 3]
```diff
@@ -131,7 +132,7 @@ def handle_result(url_info, record_info, error_info=None, http_info=None):
   pattern = settings.ignore_url_p(url_info['url'], record_info)
 
   if pattern:
-    maybe_log_ignore(url_info['url'], pattern)
+    maybe_log_ignore(url_info['url'], pattern, 'H')
```
Comment (yipdw) [inline pipeline/wpull_hooks.py]: Similar comments here: `H` -> `handle_result`.

File: pipeline/wpull_hooks.py (hunk 1 above)
Comment (yipdw) [inline pipeline/wpull_hooks.py]: > All call sites of maybe_log_ignore can be found, so no default is necessary

The `maybe_log_ignore` calls in `handle_pre_response` need to be adjusted also.

File: dashboard/dashboard.html [hunk 4]
```diff
@@ -980,6 +980,7 @@
 	logSegment.appendChild(h("div", attrs, [
 		h('span', null, " IGNOR "),
 		h('a', {"href": data["url"], "className": "ignore"}, data["url"]),
+		data["source"],
```
Comment (yipdw) [inline dashboard/dashboard.html]: I think I prefer this modification instead:

```diff
        var attrs = Reusable.obj_className_line_ignore;
        logSegment.appendChild(h("div", attrs, [
                h('span', null, " IGNOR "),
+               "(" + data["source"] + "): ",
                h('a', {"href": data["url"], "className": "ignore"}, data["url"]),
-               data["source"],
                h('span', Reusable.obj_className_bold, " by "),
                data["pattern"]
        ]));
```

This produces:

tumblrignore1

The diff in this PR requires you to horizontally scroll and has no whitespace between the URL and the source:

tumblrignore2

Comment (yipdw) [inline dashboard/dashboard.html]: Also, a question for @ivan: is this the intended use of strings in `appendChild`?  It looks like it is, but I haven't gone too far in-depth on the dashboard code.

**Comment by JesseWeinstein** [comment]
I agree with all those comments; I've now applied the specific ones.

File: dashboard/dashboard.html (hunk 4 above)
Comment (ivan) [inline dashboard/dashboard.html]: Yeah, you can use strings in the `h` children which will call `appendAny` and convert them to text nodes.

But `"(" + data["source"] + "): "` might as well be in the IGNOR span to reduce the amount of DOM.

I don't think there's any opportunity for XSS.

**Pull request merged**